### ISSUE TITLE
CNC CO2 Laser - Delay after "M" and "S" Commands

### ISSUE THREAD

**Issue by jtronics** [synthetos/TinyG#114]
Hello,
I'm using the TinyG hardware with my 40W co2 laser cutter (made in china :-) ).
For toggling the Laser I tried to use the "M3" and "M5" and for PWM the "S" commands.

It looks like:
M5 
G0 Xxxxx Yyyyy
M3
G1 Xxxxxx Yyyyyy
....

or 

M3 
S0
G0 Xxxxx Yyyyy
S100
G1 Xxxxxx Yyyyyy
S0
G0 Xxxxx Yyyyy
...
M5

The problem is, TinyG stops at every of these commands. That means, the laser decelerate to zero, than executes the M or S command and than accelerate again.

Is it possible, to modify TinyG, so that the controller executs the M or S commands without changing the speed?

There is also a video on youtube.
http://youtu.be/-CjDW2fq-C0

best regards 
Martin

**Comment by jtronics** [comment]
Hi,
my laser runs with the maximum speed (like your tests with the makerbot).
It seems, that the "M" and "S" commands are specialized handled.
Before such a command the controller turns the speed to zero, executes the command and then accelerates again. If you use the controller in a milling machine, it doesn't matter, but if you use it for a laser cutter, the resulting delay (in cause of decelerating and accelerating) is very bad.

Is it possible to change the stepping algorithm, to execute the commands, without changing the speed.

PS.: If I remove the laser command ("M" and "S") the controller runs without problems and in the expected speed.

**Comment by aldenhart** [comment]
@jtronics - Thanks. This will be changed to work as you suggest. I do not have a time frame yet, so I will leave this issue open.

**Comment by ZackSchilling** [comment]
I am glad to see this is on the list of requested features. I am currently building an automated CMM using the TinyG, a strobe light, and a camera. It would be great to be able to use the spindle or coolant signals to trigger the camera without having the machine come to a stop at each point.

**Comment by jtronics** [comment]
Hello aldenhart,
is something planned for a new release?
What do I have to change to add this feature?

**Comment by stg** [comment]
For G2 there is a TODO at plan_line.cpp:272 which I believe relates to this.
I am also quite interested in finding a way to make this work and any hints to what is necessary would be much appreciated.

**Comment by kingjamez** [comment]
Did this issue get resolved? I too would like to use a G2 to control a co2 laser.

-Jim

**Comment by interstellarspace** [comment]
Anyone tried the smoothieboard? I wonder if it has the same problem in engraving raster image

**Comment by kingjamez** [comment]
Hi interstellarspace: Yes, because of this issue, I used a smoothieboard instead of my G2 setup. I love the motion of the G2, but the smoothie won because of the laser module. The laser module is specifically built to rapidly change the laser intensity. I use Pic Engrave Pro to convert images to GCode and the smoothie will raster wonderfully and very quickly. I hear a native raster functionality is coming to smoothie soon. 

On a side note. I wish the smoothie team and the TinyG team would collaborate with their open source code to get the best of both worlds. A hybrid of the two would be a-maze-ing. 

-Jim

**Comment by interstellarspace** [comment]
Thanks for the info Jim. Where did you order it? I read some people complaining about Uberclock.

**Comment by troisiemetype** [comment]
Hi there!

Is there something new on this issue? I'm planning to buy/built a laser cutter, and would really like to use TinyG on it.
Or, can you point out where to look in the code to change this behavior? After having spent a few hours reading the code, it seems that there is some part of the work done (with moveType in planner.h), but this seems to be not implemented yet.
Looking at the code, I believe that a new "mp_queue_command" function should be made that set the MOVE_TYPE_SPINDLE_SPEED instead of the MOVE_TYPE_COMMAND. And if I understand well, the stepper.c should have a way to handle this type of action by letting it in the buffer to process it right in time, but without taking into account for the move planner.
After having understod all that by myself I'm stuck: even if I understand more or less how the planner buffer works, when looking precisely I don't find where this modification should be made...

Thank you for you help, or answer, or anything else!

**Comment by hatchethand** [comment]
I am going to revive this thread, as well (troisiemetype referenced it in #156). I have verified that the addition of S commands halts G1 movement, regardless of positioning in the file. Both of the below scenarios produce jerky raster movements:
G1X10Y20S30
        and
S30
G1X10Y20

I haven't actually checked to see if the same if true for M commands, as I'm only using M3 and M5 at the beginning and end of jobs, and S0 to turn the laser off during the job, but I see the above post from jtronics stating this to be the case for both M and S commands.

For reference, I generated my test gcode for rastering using LaserWeb.

Thanks, guys, for the awesome work!

**Comment by nickw89509** [comment]
Hi, I have made this mod for the grbl code base. I am planning to update the tinyg g2 using the same mod. Expect to have this forked for the g2 repository by mid October.

The reason this works this way is because when the planner encounters a spindle change previous GCode blocks may or may not be executing so processing of the planner que must stop so as not to change the spindle speed for previous G codes. What change involved is planning spindle changes in the same way that axes position is planned.

If you want to review the grbl modifications
https://github.com/nickw89509/LaserInk

**Comment by patyork** [comment]
No movement on this, I see.

From what I can see, the M and S commands are implemented per the comment [here](https://github.com/patyork/TinyG/blob/master/firmware/tinyg/plan_line.c#L377-L379) - with entry/exit vmax of 0 (which is a quick/short stop).

Are there any large hurdles to changing this (more than likely, optionally via a config item) to maintain velocity?

**Comment by nickw89509** [comment]
Hi Pat,
I have got sucked into a distressed project as a consultant and have been super busy for over 3 months and not able to complete this. There are two issues yes it should be easy to have the spindle command cause an instantaneous change without stopping motion. The problem is that the spindle command needs to act much like an additional axis. The Spindle speed needs to be planned just like an axis.  
Consider
G1 x1y1 S255
G1 x10y1 S225
G1 x20y1 S200
G1 x30y1 S175
G1 x40y1 S150
G1 x50y1 S125
G1 x60y1 S100
G1 x70y1 S75
G1 x80y1 S50

The planner accepts commands prior to there actual execution so if we allowed the spindle change to happen immediately then we could potentially override the desired laser intensity where the head is moving between x=10&x=20    with the power of "G1 x80y1 S50" 

In the grbl code I needed to add the spindle velocity directly to the planning block when the GCode is read. As I process each block I then send the velocity out this guarantees that the position matches the desired laser intensity which is important for image processing. 

This work quite will when velocity changes are very small between blocks or acceleration is very high. Edward has extended this algorithm to proportionally  scale the laser intensity as the instrument accelerates to full speed. This prevents over-burn when cornering as in doing fonts.

You may be interested in checking out this video
https://www.facebook.com/myobjectworks/videos/1746081008996566/
This using a streaming algorithm to send image data to a laser controlled with a custom derivative of grbl. In the video you will notice the large amount of data being sent as blocks in the software. In this video the laser is moving at about 3200 mm/min any more then that and the laser is not powerfull enough. I have tested the firmware by using it to run a CO2 laser and was able to see speeds at 15000mm/min. Note this is a totally different approach then scheduling spindle changes.
I hope this helps I would love to see this in the TinyG controllers.
Nick

**Comment by nickw89509** [comment]
replied inside of git,
cheers

**Comment by patyork** [comment]
Hi Nick,

No worries.

I've only taken a cursory glance through the code, so I'm not sure if the following will make sense.

My thought was that the Spindle (or rather, laser - as the 2 should probably not be interchangeable in this delay regard), as you said, would act similarly to an axis. However, the "moves" on this axis would be considered to be of 0 length, and can handle a jerk/accel/velocity of infinity (or to have the same velocity as the machine is currently moving in). That is, the "S" commands in this scenario would still be synchronous, but no decel would need to occur (as the "axis" can handle anything instantaneously, and is fine to continue moving at the currently velocity).

Again, not sure that the above makes sense in practice. In theory it would be like the below:

```
G0 X0
G0 X10
S2000
G0 X20
```

Currently, the planner sees an "S" code, which requires a decel to 0. However, if you image that "S" being an axis "S" that moves instantaneously, there is no need to decel. Perhaps not the best analogy (as instantaneous is not something that tends to happen in the real world), but the idea of a synchronous command that planner can essentially ignore for movement planning is what I'm getting at.

**Comment by aldenhart** [comment]
There is movement on this, it's just not ready for release. In the mean time the ideas you documented above can provide some work-arounds if you want to implement them.

**Comment by patyork** [comment]
I've tried to implement something, but without a programmer for debugging, I'm flying completely blind..

Any hints at all? Based on the comment that length, entry/exit_vmax are (or should be 0) for this type of move_type, I've tried manually changing those values before forward/backward planning, but I can't even tell if it's getting to the code block I'm expecting it to for an S command.

**Comment by patyork** [comment]
Well, a dozen or so hours stepping through the Atmel simulator debugger, I've got a workable solution. A little iffy on the first move (since it's not replannable) but a dummy move to start off the gcode and it works as expected - no decelerating into/accelerating out of an S command.

Thanks for the help.

**Comment by nickw89509** [comment]
This is awesome any have you cloned the git.
If so I would love to give this a try.

**Comment by patyork** [comment]
Yeah, I pushed it to master on [my fork](https://github.com/patyork/TinyG).

I cleaned up the code a bit, haven't retested, but it should still work. I left the version the same and pushed the .hex file as well, so you could just drop it on the board.

**Comment by patyork** [comment]
@nickw89509 I've made a pretty sizable mistake somewhere; it works fine when moving in the positive direction along an axis, but not when moving in the negative direction. I had thought that the directional part of the velocity was handled elsewhere, but it appears not.

**Comment by nickw89509** [comment]
Hmm, haven't had a chance to look at this super busy with work. When I get a free moment this is on my to do list.

**Comment by patyork** [comment]
Made a quick update today - it was the directional vector causing issues (because the spindle speed change is stored within this directional vector, as a memory optimization I guess).

Video of it in action (apologies for the simplicity of the example - if you can't tell, this was exactly my 4th test image):

[![image](https://cloud.githubusercontent.com/assets/1304633/22609120/cb4db00e-ea14-11e6-8226-27f0edb5b621.png)](https://www.youtube.com/watch?v=mprc9WvEJ6s)

[https://www.youtube.com/watch?v=mprc9WvEJ6s](https://www.youtube.com/watch?v=mprc9WvEJ6s)

**Comment by JonathanGoldnauVogt** [comment]
I'd like to see this feature added as well. Ran into this problem

**Comment by Mark-81** [comment]
As far as I understand there's still no solution in the original code, right? 
@patyork to use you firmware it's enough to update the hex file?

**Comment by Mark-81** [comment]
Ok, after flashing the @patyork's fork the behavior is slightly better, but there are still vibrations and I cannot find a suitable CAM to produce the expected paths.

I'm going to migrate to GRBL if there are no plan to add the "laser mode" to TinyG. 
By the way, neither G2 has this feature?

**Comment by patyork** [comment]
The firmware changes work well at about 2000 changes ("S" commands) per minute; this equates to my laser's abilities of a .2mm by .2mm approximate burn area at its focal point and an acceptable, if slow, machine rate of 400mm/min. The firmware begins to cause noticeable vibration (due to motors turning off for a few milliseconds on each S call) at around 5000 changes/minute, which is 1000mm/min on my machine.
I've said before, but all of my CAM software is handwritten, though some person's have fiddled other CAMs to output S command-based gcode for lasers as well.
null

**Comment by Mark-81** [comment]
Yeah, you got the point! I was machining exactly at 1000 mm / min! I'm going to give another try reducing the speed and the laser power according. Thanks for pointing it out.

Most CAM I've tried don't turn off the laser when outside of the picture but keep it at minimum level you set. Hence you will burn a dot when invert the direction.

**Comment by Mark-81** [comment]
There should be something else wrong in my setup because even at 400 mm / min I experience vibration (i.e. the motors try to stop a while instead of go ahead at current speed on consecutive G1 commands).

**Comment by ril3y** [comment]
You can try adjusting your jerk settings.  But I am not sure I follow
exactly what you are describing.  Try lowering your jerk values and see.

Riley

**Comment by Mark-81** [comment]
Good hint @ril3y but I have jerks set to an high value just to verify if the firmware works well. Anyway I will try lowering them as soon as possible and write back the results. 

Jerks are involved only when accelerating/decelerating. The expected behavior is to avoid ANY changes of speed among consecutive G1. Example:

```
F400
G1 X0.000 S100
G1 X0.200 S100
G1 X0.400 S120
G1 X0.600 S140
...
G1 X100.000 S200
```

should lead to a single G1 movement from X0 to X100, like (G1 X100.000). In the meanwhile it has just to change the PWM as soon as the related positions are reached. This is what GRBL or any other laser-oriented controller does in "laser mode".

I'm studying the code to see if I might contribute in any way (but if after 3 years there isn't an official solution yet, I'm afraid I can do anything reasonable).

**Comment by patyork** [comment]
You're right on the jerk settings; shouldn't affect anything with my changes.
Try a larger move, something like:
G1X0G1X100 S100G1X200
If the machine goes through that single S change fine (without decelerating to 0 at 100mm) then my changes are correctly flashed.
If that works fine, or if you can noticeably hear or see the machine stutter there, you'll have to adjust your feed speed.
In updated firmware, the PWM change is NOT an asynchronous change - the motors power off (I beleive) for the few milliseconds that it takes to send off the PWM change. At a certain point this is inconsequential and the machine retains its positioning accuracy and velocity just fine, but at the point (speed, quality of motors, rigidity of the machine) that you can hear or see a stutter, the brief power breaks to the motors are too frequent for the machine, and the positional accuracy and velocity can no longer be trusted as accurate.
null
